src/tokenizedSubaccounts/PPTSA.sol

```diff
@@ -0,0 +1,490 @@
+        || newParams.minMarkValueToStrikeDiffRatio > newParams.maxMarkValueToStrikeDiffRatio
+        || newParams.maxMarkValueToStrikeDiffRatio > 1e18 || newParams.maxMarkValueToStrikeDiffRatio < 1e16
+        || newParams.strikeDiff > 2e21 || newParams.strikeDiff < 1e20
+        || newParams.baseParams.optionMaxTimeToExpiry <= newParams.baseParams.optionMinTimeToExpiry
+        || newParams.baseParams.feeFactor > 0.05e18 || newParams.maxTotalCostTolerance < 2e17
+        || newParams.maxTotalCostTolerance > 5e18 || newParams.maxBuyPctOfTVL < 1e14 || newParams.maxBuyPctOfTVL > 1e18
+        || newParams.negMaxCashTolerance < 1e16 || newParams.negMaxCashTolerance > 1e18
+    ) {
+      revert PPT_InvalidParams();
+    }
+    _getPPTSAStorage().ppParams = newParams;
+
+    emit PPTSAParamsSet(newParams);
+  }
+
+  function _getBaseCollateralManagementParams()
+    internal
+    view
+    override
+    returns (BaseCollateralManagementParams storage $)
+  {
+    return _getPPTSAStorage().ppParams.baseParams;
+  }
+
+  function _getBaseCollateralManagementAddresses()
+    internal
+    view
+    override
+    returns (BaseCollateralManagementAddresses memory $)
+  {
+    return _getPPTSAStorage().baseCollateralAddresses;
+  }
+
+  ///////////////////////
+  // Action Validation //
+  ///////////////////////
+  function _verifyAction(IMatching.Action memory action, bytes32 actionHash, bytes memory extraData)
+    internal
+    virtual
+    override
+  {
+    PPTSAStorage storage $ = _getPPTSAStorage();
+
+    if (
+      action.expiry < block.timestamp + $.ppParams.minSignatureExpiry
+        || action.expiry > block.timestamp + $.ppParams.maxSignatureExpiry
+    ) {
+      revert PPT_InvalidActionExpiry();
+    }
+
+    _revokeSignature($.lastSeenHash);
+    $.lastSeenHash = actionHash;
+
+    BaseTSAAddresses memory tsaAddresses = getBaseTSAAddresses();
+
+    if (address(action.module) == address($.depositModule)) {
+      _verifyDepositAction(action, tsaAddresses);
+    } else if (address(action.module) == address($.withdrawalModule)) {
+      _verifyWithdrawAction(action, tsaAddresses);
+    } else if (address(action.module) == address($.tradeModule)) {
+      _verifyTradeAction(action, tsaAddresses);
+    } else if (address(action.module) == address($.rfqModule)) {
+      _verifyRfqAction(action, extraData);
+    } else {
+      revert PPT_InvalidModule();
+    }
+  }
+
+  /////////////////
+  // Withdrawals //
+  /////////////////
+
+  function _verifyWithdrawAction(IMatching.Action memory action, BaseTSAAddresses memory tsaAddresses) internal view {
+    IWithdrawalModule.WithdrawalData memory withdrawalData = abi.decode(action.data, (IWithdrawalModule.WithdrawalData));
+
+    if (withdrawalData.asset != address(tsaAddresses.wrappedDepositAsset)) {
+      revert PPT_InvalidAsset();
+    }
+
+    (, uint baseBalance, int cashBalance) = _getSubAccountStats();
```

Comment:
going to be adding another validation that the first parameter (left blank here) should be 0. Means we dont have any open spreads.